design-documents/node-static-content-deploy.md

```diff
@@ -0,0 +1,95 @@
+  * [Why rewrite Static Content Deployment?](#why-rewrite-static-content-deployment-)
+    + [Performance](#performance)
+    + [Access to Modern Tooling](#access-to-modern-tooling)
+    + [Accessible for Front-End Devs](#accessible-for-front-end-devs)
+    + [What opportunities open up with a re-write of SCD using node.js?](#what-opportunities-open-up-with-a-re-write-of-scd-using-nodejs-)
+  * [How do you re-write Static Content Deployment?](#how-do-you-re-write-static-content-deployment-)
+  
+## Docs
+- [Deploying Static View Files](https://devdocs.magento.com/guides/v2.3/config-guide/cli/config-cli-subcommands-static-view.html)
+- [Strategies](https://devdocs.magento.com/guides/v2.3/config-guide/cli/config-cli-subcommands-static-deploy-strategies.html)
+
+## Use Cases
+1. Pre-processing and flattening of static assets in local development (i.e. less)
+1. Pre-processing and flattening of static assets for production deployment
+
+## What does it do?
+
+- [Theme/module/i18n/"lib/web" file inheritance flattening](https://devdocs.magento.com/guides/v2.3/frontend-dev-guide/themes/theme-inherit.html)
+- Locale splitting
+    - Each locale gets its own static assets under a theme dir
+- Less compilation
+- HTML minification
+- JS minification
+- JS concatenation
+- `js-translation.json` generation
+- `requirejs-config.js` generation
+- Asset signing/versioning
+
+## Problems with Static Content Deployment
+
+There are a number of issues with the current implementation of static content deployment.
+
+1. It's prohibitively slow, sometimes taking 10+ minutes just to copy and process files
+1. There are 3 separate deployment strategies, all with different trade-offs that are not ideal. Having 3 separate ways to accomplish one task had led to [an explosion of open SCD bug reports (600+)](https://github.com/magento/magento2/search?q=static+content+deploy+is%3Aissue&type=Issues), where every strategy does not quite work as expected.
+    - **Quick**: Duplicates a ton of files, which makes cloud deployments slower
+    - **Standard**: Does way more work than it needs to, and duplicates tons of files
+    - **Compact**: Basically re-implements symlinks in the application layer, ships 150kb+ of extra JS to the front-end
+1. When something fails, it's very difficult to figure out why
+1. Lack of reliable cache + incremental builds (most people typically run it from scratch)
+1. Unable to introduce modern FE tooling (because it all expects a `node.js` runtime)
+1. Lacking granular reporting (unclear which parts of the process cause spikes in deployment times)
+
+## Why rewrite Static Content Deployment?
+
+There are 3 primary motivators for re-writing the Static Content Deployment tool
+
+### Performance
+
+The performance of `bin/magento setup:static-content:deploy` is _not great_, which leads to developers being unhappy about doing front-end work.
+
+It is also a very significant bottleneck for Magento Cloud deployments.
```

Comment:
I think we need some benchmarks to compare different PL.

Reply:
Do we? The document states multiple reasons for using node - perf is 1 of several

Reply:
I believe you may have also missed this section in the document:
--------
### Accessible for Front-End Devs
When building something new, it's important to choose the right programming language for the job.

 Often, the "right" language is based on the skillset of those that will be maintaining the tool. As an example, the Magento CLI is written using PHP. PHP is not a language designed with the intent of being used to write CLIs, but it is capable of it. Magento chose to write `bin/magento` in PHP because it makes contribution accessible to those using the tool.

 A re-write in JavaScript means that the code for static content deployment can be maintained by front-end developers, who spend most of their time in Magento working on JavaScript/HTML/CSS.

Reply:
In other words, benchmarks or not, this one point should not be the dealbreaker for this proposal.

Reply:
Yes, I have seen this section. That's why I raised this question about benchmarks. I'm not forcing to use some specific language, my point is the usage applicable language especially if we are going to introduce a new stack dependency.

Reply:
> I'm not forcing to use some specific language, my point is the usage applicable language especially if we are going to introduce a new stack dependency.

Yep! Definitely agree, and glad you're bringing it up. If we were just shooting for raw performance here, I'd definitely agree we should just choose whichever language + runtime gives us the best running times.

But, one of the critical parts of the goal here is developer experience. Leveraging the ecosystem of common tools for front-end work will allow front-end devs in the Magento world to re-apply their skills they've learned elsewhere, and get productive quicker.